### PR TITLE
initial implementation of PartitionedDataset

### PR DESCRIPTION
**Pull request by fschlimb**
# Partitioned Dataset Proposal
Alternative to #26547

I know you prefer opening an issue first, but since I had something implemented already I thought it's easier to discuss with actual code attached.

## Problem to be solved
Currently `DistributedSampler` assumes all data gets replicated on all ranks. Datasets can be too large to fit into memory. This proposal suggests a way to handle partitioned and distributed data.

## Proposed solution
This introduces a new Dataset class `PartitionedDataset` and extends `DistributedSampler` to support such datasets. All changes maintain BC. Only 2 new methods are added in `PartitionedDataset` and a single new argument is added to `DistributedSampler` for clarity.

### General idea
The major difficulty with partitioned and distributed datasets is maintaining randomness during sampling. The most general case  will require a distributed data-shuffle. The new abstract class `PartitionedDataset` provides a callback to let the implementation do exactly this: it accepts the shuffled list of indices (e.g. from `DistributedSampler`) and shuffles the data globally, which will include communication. It's up to the implementation to shuffle-data phyiscally in-place  or to keep an extra copy to store the shuffled data (and just mimic in-place operation to the outside). In the latter case the memory footprint will apparently double but for larger setups that can be still much less than the full dataset replication. 

To maintain BC we only add methods on top of what's defined in `Dataset`. Consequently, `PartitionedDataset.__len__` must return the local (per rank) data-size. Similarly, `PartitionedDataset.__getitem__` expects rank-local indices. This allows all existing uses of `Dataset` to continue working as before.  In order to allow `DistributedSampler` to shuffle all indices `PartitionedDataset` also exposes the global dataset-size through `global_len(self)`. Hence, both new methods work on the global index-space, while - as before - existing methods work on local indices.

### Discussion
* As @cpuhrsch noted hierarchical sampling with 'fixed' chunks per rank are already easily possible.
* `PartitionedDataset.shuffle_inplace` seems to also allow for supporting `IterableDataset`. As discussed in #26547 and related it not clear how useful this is, though.
* I added the argument `num_ranks` to `DistributedSampler` because `num_replicas` would be confusing. It is thinkable that `num_replicas` would get dropped/deprecated in favor of `num_ranks`.
* As suggested in #26547 this does not need any specific reader
* I have an MNIST example which patches torchvision to read the data and creates a true PGAS array (HeAT: https://github.com/helmholtz-analytics/heat). Not sure if this would be a good candidate to be included here. 

### Possible extensions
* Helper functions for global shuffling (relatively simple with `all_to_all`)
* Add a generic implementation importing the `__partitioned__´ API (https://github.com/IntelPython/DPPY-Spec/blob/draft/partitioned/Partitioned.md, https://github.com/IntelPython/DPPY-Spec/issues/3).

@thiagocrepaldi @fmassa @vincentqb @zhangguanheng66 @apaszke @csdingbin @Jian-Zhang @coquelin77 @ClaudiaComito

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### CODE REVIEW

**Comment by fschlimb** [comment]
Any thoughts, comments?

**Comment by fschlimb** [comment]
There is also an effort for defining a standard data [interchange protocol for distributed-partitioned data](https://github.com/data-apis/consortium-feedback/issues/7). It'd be nice to have pytorch in the discussion. This PR might be a basis for supporting a protocol like this.

I'd really appreciate any feedback, any thoughts or other comments.

**Comment by VitalyFedyunin** [comment]
Hi! This PR is orthogonal to the DataPipes ( #49440 and [pytorch/data](https://github.com/pytorch/data)), but I see that  PartitionedDataset can be easily translated to sharded DataPipe. I would rather avoid landing this code (unless we got a lot of requests), but I would be happy to collaborate on making this concepts available in new DataLoader and TorchData.

**Comment by fschlimb** [comment]
Thanks @VitalyFedyunin for the hint. I'll dig into DataPipes and the links you provided. As I said, my actual goal is to enable a  interchange protocol like https://github.com/data-apis/consortium-feedback/issues/7. It definitely sounds reasonable to consider the new infrastructure. I'll dig into DataPipes and the links you provided and get back to you.

**Comment by fschlimb** [comment]
@VitalyFedyunin Can I assume using DataPipes does not change how samplers are called? E.g. they are called by the DataLoader which only sees one end of the pipeline?

One of the main goals of this here, e.g. allowing global shuffling inplace, seems to contradict the overall design of datapipes. This is because datapipes deliberately decouple stages from each other, forcing data copies in each stage.

I don't think adding something like '__setitem__' to the datapipe protocol would help, even if it supported ranges.

It seems as if datapipes would need a mechanism to control the beginning of the pipeline from the sampler (something like '__shuffle__'). Which would be a mechanism/protocol similar to what is proposed here.

Any thoughts?
